sdk/communication/AzureCommunicationChat/Source/Registrar/RegistrarClient.swift

```diff
@@ -0,0 +1,267 @@
+    }
+
+    /// Sets the header on a Registrar request.
+    /// - Parameters:
+    ///   - request: The  HTTP request.
+    ///   - completionHandler: Returns the request, or an error if the request header failed to be created or added
+    private func setRequestHeader(
+        on request: HTTPRequest,
+        completionHandler: @escaping (HTTPRequest?, AzureError?) -> Void
+    ) {
+        credential.token { accessToken, error in
+            do {
+                // Get token from CommunicationTokenCredential to set the authentication header
+                guard let skypeToken = accessToken?.token else {
+                    completionHandler(
+                        nil,
+                        AzureError.client("Failed to get token from CommunicationTokenCredential.", error)
+                    )
+                    return
+                }
+
+                let httpHeaders: HTTPHeaders = [
+                    RegistrarHeader.contentType.rawValue: RegistrarMimeType.json.rawValue,
+                    RegistrarHeader.skypeTokenHeader.rawValue: skypeToken
+                ]
+
+                let httpRequest = try HTTPRequest(
+                    method: request.httpMethod,
+                    url: request.url,
+                    headers: httpHeaders,
+                    data: request.data
+                )
+
+                completionHandler(httpRequest, nil)
+            } catch {
+                let azureError = AzureError.client("Failed to add headers to the Register request.", error)
+                completionHandler(nil, azureError)
+            }
+        }
+    }
+
+    /// Sends an HTTP request to Registrar.
+    /// - Parameters:
+    ///   - request: The HTTP request.
+    ///   - completionHandler: Returns the URLResponse, and an error if any errors occurred.
+    private func sendHttpRequest(
+        _ request: HTTPRequest,
+        completionHandler: @escaping (Result<HTTPResponse?, AzureError>) -> Void
+    ) {
+        let dispatchQueue = DispatchQueue.main
+
+        let context = PipelineContext.of(keyValues: [
+            ContextKey.allowedStatusCodes.rawValue: [202] as AnyObject
+        ])
+
+        self.request(request, context: context) { result, httpResponse in
+            switch result {
+            case .success:
+                guard (httpResponse?.statusCode) == 202 else {
+                    let invalidStatusCodeError = AzureError.client("Didn't find a valid status code.")
+                    dispatchQueue.async {
+                        completionHandler(.failure(invalidStatusCodeError))
+                    }
+                    return
+                }
+                dispatchQueue.async {
+                    completionHandler(.success(httpResponse))
+                }
+            case .failure:
+                dispatchQueue.async {
+                    completionHandler(.failure(AzureError.client("Registration request failed.")))
+                }
```

Comment:
There would presumably be an error here, and by not including this inner error you are potentially throwing away useful error info and replacing it with generic, unactionable info.